File: frontend/src/instrument/trend.ts
/** Trend napříč timeframy (#1089, karta Trend v Briefingu).

Čtení shora dolů, jak to dělají diskreční tradeři: vyšší timeframe (W, D)
určuje směr, nižší (4h, 1h, 15m) načasování. Per timeframe se kombinují dvě
nezávislé věci:

- **Struktura trhu** — fraktálové pivoty (swing high = svíčka s nejvyšším high
  v okně ±n, swing low obdobně). Poslední dva swing highs i lows rostou
  = HH/HL = rostoucí; klesají = LH/LL = klesající; jinak bez trendu.
- **Klouzavé průměry** — EMA20 a EMA50: cena vs. EMA20, EMA20 vs. EMA50, sklon
  EMA20 za posledních 5 svíček.

Verdikt timeframe je silný, když struktura i EMA souhlasí; slabý, když směr
říká jen jedna z nich. Bez dostatku svíček se nic nedosazuje — řádek říká
„málo dat" (stejný princip jako ADR-0028).

Čisté funkce bez React závislostí — testují se samostatně. */

export interface Candle {
  ts: string
  open: number
  high: number
  low: number
  close: number
  volume: number
  /** Rozdělaná svíčka (API): struktura ji nepoužívá, EMA ano (průběžný stav). */
  partial?: boolean
}

export type TrendDirection = 'up' | 'down' | 'range'

export const TIMEFRAMES = ['W', 'D', '240', '60', '15'] as const
export type TimeframeKey = (typeof TIMEFRAMES)[number]

export const TIMEFRAME_LABELS: Record<TimeframeKey, string> = {
  W: 'Týden',
  D: 'Den',
  '240': '4h',
  '60': '1h',
  '15': '15m',
}

/** Vyšší timeframy určují směr, nižší načasování. */
export const HIGHER_TIMEFRAMES: readonly TimeframeKey[] = ['W', 'D']
export const LOWER_TIMEFRAMES: readonly TimeframeKey[] = ['240', '60', '15']

/** Šířka fraktálu: D/W mají méně svíček, užší okno; intradenní širší proti šumu. */
const PIVOT_WIDTH: Record<TimeframeKey, number> = { W: 2, D: 3, '240': 3, '60': 3, '15': 3 }

/** Minimum svíček pro verdikt — EMA20 potřebuje aspoň své okno. */
export const MIN_CANDLES = 20
const EMA_FAST = 20
const EMA_SLOW = 50
const SLOPE_LOOKBACK = 5

export interface Pivot {
  index: number
  price: number
  kind: 'high' | 'low'
}

export type StructureKind = 'HH/HL' | 'LH/LL' | 'mixed'

export interface StructureReading {
  kind: StructureKind
  lastHigh: number | null
  lastLow: number | null
}

export interface EmaReading {
  ema20: number
  /** Null, když je svíček méně než 50. */
  ema50: number | null
  priceAboveFast: boolean
  /** Null bez EMA50. */
  fastAboveSlow: boolean | null
  /** Sklon EMA20 za posledních 5 svíček v bodech. */
  slope: number
  direction: TrendDirection
}

export interface TimeframeTrend {
  tf: TimeframeKey
  candles: number
  direction: TrendDirection | null
  strength: 'strong' | 'weak' | null
  structure: StructureReading | null
  ema: EmaReading | null
  /** Krátké vysvětlení pro tabulku; „málo dat" bez verdiktu. */
  note: string
}

export interface TrendReport {
  byTimeframe: TimeframeTrend[]
  /** Směr vyšších TF (W, D); null = ani jeden nemá verdikt. */
  higher: TrendDirection | null
  /** Směr nižších TF (4h, 1h, 15m) — většina; null bez verdiktu. */
  lower: TrendDirection | null
  /** Kolik TF s verdiktem souhlasí se směrem vyšších TF / kolik jich verdikt má. */
  aligned: number
  decided: number
  /** Čtení pro den jednou až dvěma větami. */
  reading: string
  /** Očekávaný směr z trendu pro verdikt dne (#1090): null = bez převahy. */
  expected: TrendDirection | null
}

/** Exponenciální klouzavý průměr; první hodnota = prostý průměr prvních `period`. */
export function ema(values: number[], period: number): number[] {
  if (values.length < period) return []
  const k = 2 / (period + 1)
  const seed = values.slice(0, period).reduce((sum, value) => sum + value, 0) / period
  const result: number[] = [seed]
  for (let i = period; i < values.length; i += 1) {
    const previous = result[result.length - 1]
    result.push(values[i] * k + previous * (1 - k))
  }
  return result
}

/** Fraktálové pivoty: high vyšší než `width` sousedů na obě strany (low obdobně). */
export function findPivots(candles: Candle[], width: number): Pivot[] {
  const pivots: Pivot[] = []
  for (let i = width; i < candles.length - width; i += 1) {
    let isHigh = true
    let isLow = true
    for (let offset = 1; offset <= width; offset += 1) {
      const left = candles[i - offset]
      const right = candles[i + offset]
      if (candles[i].high <= left.high || candles[i].high <= right.high) isHigh = false
      if (candles[i].low >= left.low || candles[i].low >= right.low) isLow = false
      if (!isHigh && !isLow) break
    }
    if (isHigh) pivots.push({ index: i, price: candles[i].high, kind: 'high' })
    if (isLow) pivots.push({ index: i, price: candles[i].low, kind: 'low' })
  }
  return pivots
}

/** Struktura z posledních dvou swing highs a lows; null bez dvou od každého. */
export function readStructure(pivots: Pivot[]): StructureReading | null {
  const highs = pivots.filter((pivot) => pivot.kind === 'high')
  const lows = pivots.filter((pivot) => pivot.kind === 'low')
  if (highs.length < 2 || lows.length < 2) return null
  const [prevHigh, lastHigh] = highs.slice(-2)
  const [prevLow, lastLow] = lows.slice(-2)
  const higherHighs = lastHigh.price > prevHigh.price
  const higherLows = lastLow.price > prevLow.price
  const lowerHighs = lastHigh.price < prevHigh.price
  const lowerLows = lastLow.price < prevLow.price
  let kind: StructureKind = 'mixed'
  if (higherHighs && higherLows) kind = 'HH/HL'
  else if (lowerHighs && lowerLows) kind = 'LH/LL'
  return { kind, lastHigh: lastHigh.price, lastLow: lastLow.price }
}

/** EMA čtení z close; null pod MIN_CANDLES. */
export function readEma(candles: Candle[]): EmaReading | null {
  if (candles.length < MIN_CANDLES) return null
  const closes = candles.map((candle) => candle.close)
  const fast = ema(closes, EMA_FAST)
  const slow = closes.length >= EMA_SLOW ? ema(closes, EMA_SLOW) : []
  const ema20 = fast[fast.length - 1]
  const ema50 = slow.length > 0 ? slow[slow.length - 1] : null
  const lookback = Math.min(SLOPE_LOOKBACK, fast.length - 1)
  const slope = ema20 - fast[fast.length - 1 - lookback]
  const price = closes[closes.length - 1]
  const priceAboveFast = price > ema20
  const fastAboveSlow = ema50 === null ? null : ema20 > ema50
  let direction: TrendDirection = 'range'
  if (priceAboveFast && slope > 0 && fastAboveSlow !== false) direction = 'up'
  else if (!priceAboveFast && slope < 0 && fastAboveSlow !== true) direction = 'down'
  return { ema20, ema50, priceAboveFast, fastAboveSlow, slope, direction }
}

function structureDirection(structure: StructureReading | null): TrendDirection | null {
  if (structure === null) return null
  if (structure.kind === 'HH/HL') return 'up'
  if (structure.kind === 'LH/LL') return 'down'
  return 'range'
}

const DIRECTION_LABELS: Record<TrendDirection, string> = {
  up: 'rostoucí',
  down: 'klesající',
  range: 'bez trendu',
}

export function directionLabel(direction: TrendDirection | null): string {
  return direction === null ? '—' : DIRECTION_LABELS[direction]
}

/** Verdikt jednoho timeframe. Struktura se čte jen z uzavřených svíček. */
export function assessTimeframe(tf: TimeframeKey, candles: Candle[]): TimeframeTrend {
  const closed = candles.filter((candle) => !candle.partial)
  if (candles.length < MIN_CANDLES) {
    return {
      tf,
      candles: candles.length,
      direction: null,
      strength: null,
      structure: null,
      ema: null,
      note: `málo dat (${candles.length}/${MIN_CANDLES} svíček)`,
    }
  }
  const structure = readStructure(findPivots(closed, PIVOT_WIDTH[tf]))
  const emaReading = readEma(candles)
  const fromStructure = structureDirection(structure)
  const fromEma = emaReading?.direction ?? null
  let direction: TrendDirection | null = null
  let strength: 'strong' | 'weak' | null = null
  if (fromStructure !== null && fromStructure !== 'range' && fromStructure === fromEma) {
    direction = fromStructure
    strength = 'strong'
  } else if (fromStructure !== null && fromStructure !== 'range' && fromEma === 'range') {
    direction = fromStructure
    strength = 'weak'
  } else if ((fromStructure === null || fromStructure === 'range') && fromEma !== null) {
    direction = fromEma
    strength = fromEma === 'range' ? null : 'weak'
  } else if (fromStructure !== null && fromEma !== null) {
    // Struktura a EMA proti sobě → bez trendu (korekce uvnitř trendu)
    direction = 'range'
    strength = null
  }
  const notes: string[] = []
  if (structure) notes.push(`struktura ${structure.kind}`)
  else notes.push('struktura bez dvou swingů')
  if (emaReading) {
    notes.push(
      `${emaReading.priceAboveFast ? 'cena nad' : 'cena pod'} EMA20` +
        (emaReading.fastAboveSlow === null
          ? ''
          : `, EMA20 ${emaReading.fastAboveSlow ? 'nad' : 'pod'} EMA50`),
    )
  }
  if (strength === 'weak') notes.push('slabý — struktura a EMA se neshodnou')
  return { tf, candles: candles.length, direction, strength, structure, ema: emaReading, note: notes.join('; ') } // prettier-ignore
}

function majority(directions: (TrendDirection | null)[]): TrendDirection | null {
  const decided = directions.filter((value): value is TrendDirection => value !== null)
  if (decided.length === 0) return null
  const counts: Record<TrendDirection, number> = { up: 0, down: 0, range: 0 }
  for (const value of decided) counts[value] += 1
  if (counts.up > counts.down && counts.up >= counts.range) return 'up'
  if (counts.down > counts.up && counts.down >= counts.range) return 'down'
  return 'range'
}

/** Směr vyšších TF: shoda W a D, jinak rozhoduje D (W je kontext, D obchodní rámec). */
function higherDirection(byTf: Map<TimeframeKey, TimeframeTrend>): TrendDirection | null {
  const week = byTf.get('W')?.direction ?? null
  const day = byTf.get('D')?.direction ?? null
  if (day === null) return week
  return day
}

/** Sestaví report shora dolů; chybějící TF (bez dat z API) se přeskočí. */
export function assessTrends(candlesByTf: Partial<Record<TimeframeKey, Candle[]>>): TrendReport {
  const byTimeframe = TIMEFRAMES.map((tf) => assessTimeframe(tf, candlesByTf[tf] ?? []))
  const map = new Map(byTimeframe.map((row) => [row.tf, row] as const))
  const higher = higherDirection(map)
  const lower = majority(LOWER_TIMEFRAMES.map((tf) => map.get(tf)?.direction ?? null))
  const decidedRows = byTimeframe.filter((row) => row.direction !== null)
  const decided = decidedRows.length
  const aligned = higher === null ? 0 : decidedRows.filter((row) => row.direction === higher).length
  const week = map.get('W')?.direction ?? null
  const day = map.get('D')?.direction ?? null
  return {
    byTimeframe,
    higher,
    lower,
    aligned,
    decided,
    reading: buildReading({ higher, lower, week, day, aligned, decided }),
    expected: expectedDirection(higher, lower),
  }
}

/** Očekávaný směr z trendu: vyšší TF rozhoduje, souhlas nižších ho potvrzuje.
Vyšší bez trendu → bez převahy. Nižší proti vyššímu → převaha zůstává u vyššího,
ale slabší (verdikt dne to řeší vahami, #1090). */
export function expectedDirection(
  higher: TrendDirection | null,
  lower: TrendDirection | null,
): TrendDirection | null {
  if (higher === null) return lower === null || lower === 'range' ? null : lower
  if (higher === 'range') return null
  return higher
}

function buildReading(input: {
  higher: TrendDirection | null
  lower: TrendDirection | null
  week: TrendDirection | null
  day: TrendDirection | null
  aligned: number
  decided: number
}): string {
  const { higher, lower, week, day, aligned, decided } = input
  if (decided === 0) return 'Zatím málo svíček pro čtení trendu.'
  const parts: string[] = []
  if (week !== null && day !== null && week !== day && day === 'range') {
    parts.push(
      `Týden ${directionLabel(week)}, den bez trendu — konsolidace uvnitř týdenního trendu.`,
    )
  } else if (week !== null && day !== null && week !== day) {
    parts.push(
      `Týden ${directionLabel(week)}, den ${directionLabel(day)} — denní trend je korekce uvnitř týdenního.`,
    )
  } else if (higher !== null) {
    parts.push(`Vyšší TF: ${directionLabel(higher)}.`)
  }
  if (higher === null || higher === 'range') {
    parts.push('Vyšší TF bez trendu — den řídí úrovně a gamma režim, ne trend.')
    return parts.join(' ')
  }
  const side = higher === 'up' ? 'long' : 'short'
  const counter = higher === 'up' ? 'short' : 'long'
  if (lower === higher) {
    parts.push(
      `Nižší TF souhlasí (${aligned}/${decided}) — obchodovat ${side} ve směru trendu, ${counter} jen jako fade u zdí.`,
    )
  } else if (lower === null || lower === 'range') {
    parts.push(
      `Nižší TF bez trendu — čekat na obnovení směru vyššího TF (${side}), do té doby úrovně.`,
    )
  } else {
    parts.push(
      `Nižší TF korigují proti vyššímu (${aligned}/${decided} v souladu) — ${side} až po návratu nižších TF do směru, ${counter} jen s menší velikostí.`,
    )
  }
  return parts.join(' ')
}

File: engine/src/gexlens_engine/compute/trend.py
"""Trend napříč timeframy — port `frontend/src/instrument/trend.ts` (#1089) pro engine.

Stejná pravidla jako karta Trend v Briefingu, ať automatický scénář (#1173 A)
hlasuje shodně s tím, co uživatel vidí: struktura z fraktálových pivotů
(HH/HL vs. LH/LL) + EMA20/EMA50 se sklonem; vyšší TF (W, D) určuje směr,
nižší (4h, 1h, 15m) načasování. Čisté funkce; svíčky přicházejí z API
`/candles/{symbol}?tf=` (tytéž, které čte frontend).
"""

from collections import Counter
from dataclasses import dataclass
from typing import Literal

Direction = Literal["up", "down", "range"]

TIMEFRAMES: tuple[str, ...] = ("W", "D", "240", "60", "15")
HIGHER_TIMEFRAMES: tuple[str, ...] = ("W", "D")
LOWER_TIMEFRAMES: tuple[str, ...] = ("240", "60", "15")
PIVOT_WIDTH: dict[str, int] = {"W": 2, "D": 3, "240": 3, "60": 3, "15": 3}
MIN_CANDLES = 20
EMA_FAST = 20
EMA_SLOW = 50
SLOPE_LOOKBACK = 5


@dataclass(frozen=True)
class Candle:
    high: float
    low: float
    close: float
    partial: bool = False


@dataclass(frozen=True)
class TimeframeTrend:
    tf: str
    candles: int
    direction: Direction | None
    strength: str | None


@dataclass(frozen=True)
class TrendReport:
    higher: Direction | None
    lower: Direction | None
    expected: Direction | None
    by_timeframe: tuple[TimeframeTrend, ...]


def ema(values: list[float], period: int) -> list[float]:
    if len(values) < period:
        return []
    k = 2 / (period + 1)
    seed = sum(values[:period]) / period
    result = [seed]
    for value in values[period:]:
        result.append(value * k + result[-1] * (1 - k))
    return result


def find_pivots(candles: list[Candle], width: int) -> list[tuple[int, float, str]]:
    pivots: list[tuple[int, float, str]] = []
    for i in range(width, len(candles) - width):
        is_high = True
        is_low = True
        for offset in range(1, width + 1):
            left = candles[i - offset]
            right = candles[i + offset]
            if candles[i].high <= left.high or candles[i].high <= right.high:
                is_high = False
            if candles[i].low >= left.low or candles[i].low >= right.low:
                is_low = False
            if not is_high and not is_low:
                break
        if is_high:
            pivots.append((i, candles[i].high, "high"))
        if is_low:
            pivots.append((i, candles[i].low, "low"))
    return pivots


def structure_direction(pivots: list[tuple[int, float, str]]) -> Direction | None:
    highs = [p for p in pivots if p[2] == "high"]
    lows = [p for p in pivots if p[2] == "low"]
    if len(highs) < 2 or len(lows) < 2:
        return None
    prev_high, last_high = highs[-2], highs[-1]
    prev_low, last_low = lows[-2], lows[-1]
    if last_high[1] > prev_high[1] and last_low[1] > prev_low[1]:
        return "up"
    if last_high[1] < prev_high[1] and last_low[1] < prev_low[1]:
        return "down"
    return "range"


def ema_direction(candles: list[Candle]) -> Direction | None:
    if len(candles) < MIN_CANDLES:
        return None
    closes = [c.close for c in candles]
    fast = ema(closes, EMA_FAST)
    slow = ema(closes, EMA_SLOW) if len(closes) >= EMA_SLOW else []
    ema20 = fast[-1]
    ema50 = slow[-1] if slow else None
    lookback = min(SLOPE_LOOKBACK, len(fast) - 1)
    slope = ema20 - fast[-1 - lookback]
    price = closes[-1]
    price_above_fast = price > ema20
    fast_above_slow = None if ema50 is None else ema20 > ema50
    if price_above_fast and slope > 0 and fast_above_slow is not False:
        return "up"
    if not price_above_fast and slope < 0 and fast_above_slow is not True:
        return "down"
    return "range"


def assess_timeframe(tf: str, candles: list[Candle]) -> TimeframeTrend:
    if len(candles) < MIN_CANDLES:
        return TimeframeTrend(tf=tf, candles=len(candles), direction=None, strength=None)
    closed = [c for c in candles if not c.partial]
    from_structure = structure_direction(find_pivots(closed, PIVOT_WIDTH.get(tf, 3)))
    from_ema = ema_direction(candles)
    direction: Direction | None = None
    strength: str | None = None
    if from_structure is not None and from_structure != "range" and from_structure == from_ema:
        direction, strength = from_structure, "strong"
    elif from_structure is not None and from_structure != "range" and from_ema == "range":
        direction, strength = from_structure, "weak"
    elif (from_structure is None or from_structure == "range") and from_ema is not None:
        direction = from_ema
        strength = None if from_ema == "range" else "weak"
    elif from_structure is not None and from_ema is not None:
        direction, strength = "range", None
    return TimeframeTrend(tf=tf, candles=len(candles), direction=direction, strength=strength)


def _majority(directions: list[Direction | None]) -> Direction | None:
    decided = [d for d in directions if d is not None]
    if not decided:
        return None
    counts = Counter(decided)
    up, down, rng = counts["up"], counts["down"], counts["range"]
    if up > down and up >= rng:
        return "up"
    if down > up and down >= rng:
        return "down"
    return "range"


def expected_direction(higher: Direction | None, lower: Direction | None) -> Direction | None:
    if higher is None:
        return None if lower is None or lower == "range" else lower
    if higher == "range":
        return None
    return higher


def assess_trends(candles_by_tf: dict[str, list[Candle]]) -> TrendReport:
    rows = tuple(assess_timeframe(tf, candles_by_tf.get(tf, [])) for tf in TIMEFRAMES)
    by_tf = {row.tf: row for row in rows}
    week = by_tf["W"].direction
    day = by_tf["D"].direction
    higher = week if day is None else day
    lower = _majority([by_tf[tf].direction for tf in LOWER_TIMEFRAMES])
    return TrendReport(
        higher=higher, lower=lower, expected=expected_direction(higher, lower), by_timeframe=rows
    )
